### PR TITLE
auto install certain external commands

### PR DESCRIPTION
**Pull request by xu-cheng**
* Replace `install_tap` with `Tap#install`.
* Auto install certain external commands, therefore it will behavior like they are pre-installed.

cc @vitorgalvao

### CODE REVIEW

File: Library/brew.rb [hunk 1]
```diff
@@ -152,6 +152,28 @@ def require?(path)
...
+    end
+
+    if possible_tap && !possible_tap.installed?
+      possible_tap.install
+
+      if cmd == "cask"
+        require "cmd/install"
+        brew_cask = Formulary.factory("brew-cask")
+        Homebrew.install_formula(brew_cask)
+      end
```
Comment (xu-cheng) [inline Library/brew.rb]: It would be prefect, if `cask` is a tap external commands rather than standalone formula.

Comment (mikemcquaid) [inline Library/brew.rb]: Agreed. One of us should probably submit a PR to cask to do that for them.

File: Library/Homebrew/cmd/tap.rb [hunk 2]
```diff
@@ -14,48 +13,22 @@ def tap
...
     else
       user, repo = tap_args
-      clone_target = ARGV.named[1]
-      opoo "#{user}/#{repo} already tapped!" unless install_tap(user, repo, clone_target)
+      tap = Tap.fetch(user, repo)
+      tap.install(:clone_target => ARGV.named[1], :full_clone => ARGV.include?("--full"))
     end
   end
 
+  # @deprecated this method will be removed in the future, if no external commands use it.
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/tap.rb]: Maybe print a warning inside the method saying that?

Comment (mikemcquaid) [inline Library/Homebrew/cmd/tap.rb]: Oh, you already did :grinning:

File: Library/brew.rb [hunk 3]
```diff
@@ -152,6 +152,28 @@ def require?(path)
...
   else
+    require "tap"
+    possible_tap = case cmd
+    when *%w[brewdle brewdler bundle bundler]
+      Tap.fetch("Homebrew", "bundle")
+    when "cask"
+      Tap.fetch("caskroom", "cask")
+    when "services"
+      Tap.fetch("Homebrew", "services")
+    end
```
Comment (mikemcquaid) [inline Library/brew.rb]: Very nice.

**Comment by mikemcquaid** [comment]
Love this :heart_eyes:

**Comment by vitorgalvao** [comment]
So as I understand it, as soon as someone `brew cask {{whatever}}`, homebrew will take care of all the dirty work of tapping and whatnot. Brilliant! I like this very much. [One less thing to worry about](https://github.com/caskroom/homebrew-cask/pull/15142/files).

Pinging the rest of @caskroom.

**Comment by xu-cheng** [comment]
> So as I understand it, as soon as someone brew cask {{whatever}}, homebrew will take care of all the dirty work of tapping and whatnot. 

That's right. 

Also as mentioned in above command, it will offer better user experience if cask is a tap external command rather than standalone formula.

**Comment by jawshooah** [comment]
This is great! One step closer to full integration :smile:

We would very much appreciate any help in transitioning from a formula to an external command.

**Comment by victorpopkov** [comment]
Amazing! Definitely makes the life simpler. Really like this feature. :+1:

**Comment by fanquake** [comment]
👍🏼

**Pull request closed without merging**